File: cpp/equity-var-engine/tests/test_matrix.cpp
// Matrix, Cholesky (clean + jitter fallback), covariance estimators.
#include <gtest/gtest.h>

#include <cmath>
#include <vector>

#include "eqvar/matrix.hpp"

using namespace eqvar;

namespace {

Matrix spd3() {
    // SPD matrix built as A = B B^T + I from a fixed B.
    return Matrix(3, 3,
                  {4.0, 2.0, 0.6,
                   2.0, 5.0, 1.5,
                   0.6, 1.5, 3.0});
}

Matrix reconstruct(const Matrix& l) {
    const std::size_t n = l.rows();
    Matrix a(n, n, 0.0);
    for (std::size_t i = 0; i < n; ++i)
        for (std::size_t j = 0; j < n; ++j) {
            double s = 0.0;
            for (std::size_t k = 0; k < n; ++k) s += l(i, k) * l(j, k);
            a(i, j) = s;
        }
    return a;
}

}  // namespace

TEST(Cholesky, ReconstructsSpdTo1e12) {
    const Matrix a = spd3();
    const CholeskyResult res = cholesky(a);
    EXPECT_DOUBLE_EQ(res.jitter_added, 0.0);
    EXPECT_LT(Matrix::max_abs_diff(reconstruct(res.lower), a), 1e-12);
    // Lower-triangular: strict upper part is zero.
    EXPECT_DOUBLE_EQ(res.lower(0, 1), 0.0);
    EXPECT_DOUBLE_EQ(res.lower(0, 2), 0.0);
    EXPECT_DOUBLE_EQ(res.lower(1, 2), 0.0);
}

TEST(Cholesky, JitterPathOnSingularMatrix) {
    // Perfectly correlated 2x2 — exactly singular, plain Cholesky must fail
    // and the diagonal-jitter fallback must engage.
    const Matrix a(2, 2, {1.0, 1.0, 1.0, 1.0});
    const CholeskyResult res = cholesky(a);
    EXPECT_GT(res.jitter_added, 0.0);
    EXPECT_LE(res.jitter_added, 1e-6);  // tiny relative to unit variances
    EXPECT_LT(Matrix::max_abs_diff(reconstruct(res.lower), a), 1e-6);
}

TEST(Cholesky, JitterPathOnZeroVarianceAsset) {
    const Matrix a(2, 2, {0.01, 0.0, 0.0, 0.0});  // second asset has zero variance
    const CholeskyResult res = cholesky(a);
    EXPECT_GT(res.jitter_added, 0.0);
    EXPECT_LT(Matrix::max_abs_diff(reconstruct(res.lower), a), 1e-8);
}

TEST(Cholesky, RejectsBadInput) {
    EXPECT_THROW(cholesky(Matrix(2, 3, 1.0)), std::invalid_argument);           // non-square
    EXPECT_THROW(cholesky(Matrix(2, 2, {1.0, 0.5, -0.5, 1.0})), std::invalid_argument);  // asymmetric
    // Badly indefinite: jitter cannot rescue a huge negative eigenvalue.
    EXPECT_THROW(cholesky(Matrix(2, 2, {1.0, 1e8, 1e8, 1.0})), std::runtime_error);
}

TEST(MatVec, KnownProductAndShapeCheck) {
    const Matrix a(2, 3, {1.0, 2.0, 3.0, 4.0, 5.0, 6.0});
    const std::vector<double> x = {1.0, 0.5, -1.0};
    const std::vector<double> y = matvec(a, x);
    ASSERT_EQ(y.size(), 2u);
    EXPECT_DOUBLE_EQ(y[0], -1.0);
    EXPECT_DOUBLE_EQ(y[1], 0.5);
    EXPECT_THROW(matvec(a, std::vector<double>{1.0, 2.0}), std::invalid_argument);
}

TEST(QuadForm, MatchesHandComputation) {
    const Matrix s(2, 2, {2.0, 0.5, 0.5, 1.0});
    const std::vector<double> w = {1.0, -2.0};
    // w'Sw = 2 - 2*0.5*2 + 4 = 4
    EXPECT_DOUBLE_EQ(quad_form(w, s), 4.0);
}

TEST(SampleCovariance, HandComputed2x2) {
    // Two assets, three days: covariance ddof=1 by hand.
    const Matrix r(3, 2, {0.01, 0.02, -0.01, 0.00, 0.03, 0.01});
    const Matrix s = sample_covariance(r);
    EXPECT_NEAR(s(0, 0), 4.0e-4, 1e-18);
    EXPECT_NEAR(s(1, 1), 1.0e-4, 1e-18);
    EXPECT_NEAR(s(0, 1), 1.0e-4, 1e-18);
    EXPECT_DOUBLE_EQ(s(0, 1), s(1, 0));
    EXPECT_THROW(sample_covariance(Matrix(1, 2, 0.0)), std::invalid_argument);
}

TEST(EwmaCovariance, SinglePointLimitAndValidation) {
    // With lam -> 1 the EWMA forecast stays near the sample seed.
    const Matrix r(3, 2, {0.01, 0.02, -0.01, 0.00, 0.03, 0.01});
    const Matrix s = sample_covariance(r);
    const Matrix e = ewma_covariance(r, 0.999999);
    EXPECT_LT(Matrix::max_abs_diff(e, s), 1e-8);
    EXPECT_THROW(ewma_covariance(r, 1.0), std::invalid_argument);
    EXPECT_THROW(ewma_covariance(r, 0.0), std::invalid_argument);
}

TEST(CovarianceFromVols, Known) {
    const std::vector<double> vols = {0.01, 0.02};
    const Matrix corr(2, 2, {1.0, 0.5, 0.5, 1.0});
    const Matrix cov = covariance_from_vols(vols, corr);
    EXPECT_DOUBLE_EQ(cov(0, 0), 1e-4);
    EXPECT_DOUBLE_EQ(cov(0, 1), 1e-4);
    EXPECT_DOUBLE_EQ(cov(1, 1), 4e-4);
}

File: rust/equity-var-engine/tests/test_matrix.rs
//! Matrix, Cholesky (clean + jitter fallback), covariance estimators.

use eq_var_engine::matrix::{covariance_from_vols, ewma_covariance, sample_covariance, Matrix};
use eq_var_engine::EqVarError;

fn spd3() -> Matrix {
    // SPD matrix built as A = B B^T + I from a fixed B.
    Matrix::from_vec(3, 3, vec![4.0, 2.0, 0.6, 2.0, 5.0, 1.5, 0.6, 1.5, 3.0]).unwrap()
}

fn reconstruct(l: &Matrix) -> Matrix {
    let n = l.rows();
    let mut a = Matrix::zeros(n, n);
    for i in 0..n {
        for j in 0..n {
            let mut s = 0.0;
            for k in 0..n {
                s += l.get(i, k) * l.get(j, k);
            }
            a.set(i, j, s);
        }
    }
    a
}

fn max_abs_diff(a: &Matrix, b: &Matrix) -> f64 {
    a.as_slice()
        .iter()
        .zip(b.as_slice().iter())
        .fold(0.0f64, |m, (x, y)| m.max((x - y).abs()))
}

#[test]
fn cholesky_reconstructs_spd() {
    let a = spd3();
    let l = a.cholesky().unwrap();
    assert!(max_abs_diff(&reconstruct(&l), &a) < 1e-12);
    assert_eq!(l.get(0, 1), 0.0);
    assert_eq!(l.get(0, 2), 0.0);
    assert_eq!(l.get(1, 2), 0.0);
}

#[test]
fn cholesky_jitter_path_on_singular_matrix() {
    // Perfectly correlated 2x2 — exactly singular, plain Cholesky must fail
    // and the diagonal-jitter fallback must engage.
    let a = Matrix::from_vec(2, 2, vec![1.0, 1.0, 1.0, 1.0]).unwrap();
    assert!(a.cholesky().is_err());
    let l = a.cholesky_jitter(1e-10, 12).unwrap();
    assert!(max_abs_diff(&reconstruct(&l), &a) < 1e-6);
}

#[test]
fn cholesky_jitter_path_on_zero_variance_asset() {
    let a = Matrix::from_vec(2, 2, vec![0.01, 0.0, 0.0, 0.0]).unwrap();
    let l = a.cholesky_jitter(1e-10, 12).unwrap();
    assert!(max_abs_diff(&reconstruct(&l), &a) < 1e-8);
}

#[test]
fn cholesky_rejects_bad_input() {
    let non_square = Matrix::from_vec(2, 3, vec![1.0; 6]).unwrap();
    assert!(matches!(non_square.cholesky(), Err(EqVarError::InvalidInput(_))));
    let asym = Matrix::from_vec(2, 2, vec![1.0, 0.5, -0.5, 1.0]).unwrap();
    assert!(matches!(asym.cholesky(), Err(EqVarError::InvalidInput(_))));
    // Badly indefinite: jitter cannot rescue a huge negative eigenvalue.
    let bad = Matrix::from_vec(2, 2, vec![1.0, 1e8, 1e8, 1.0]).unwrap();
    assert!(matches!(bad.cholesky_jitter(1e-10, 12), Err(EqVarError::Numerical(_))));
}

#[test]
fn matvec_known_product_and_shape_check() {
    let a = Matrix::from_vec(2, 3, vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0]).unwrap();
    let x = [1.0, 0.5, -1.0];
    let y = a.matvec(&x).unwrap();
    assert_eq!(y.len(), 2);
    assert_eq!(y[0], -1.0);
    assert_eq!(y[1], 0.5);
    assert!(a.matvec(&[1.0, 2.0]).is_err());
}

#[test]
fn sample_covariance_hand_computed_2x2() {
    let r = Matrix::from_vec(3, 2, vec![0.01, 0.02, -0.01, 0.00, 0.03, 0.01]).unwrap();
    let s = sample_covariance(&r).unwrap();
    assert!((s.get(0, 0) - 4.0e-4).abs() < 1e-18);
    assert!((s.get(1, 1) - 1.0e-4).abs() < 1e-18);
    assert!((s.get(0, 1) - 1.0e-4).abs() < 1e-18);
    assert_eq!(s.get(0, 1), s.get(1, 0));
    assert!(sample_covariance(&Matrix::zeros(1, 2)).is_err());
}

#[test]
fn ewma_covariance_single_point_limit_and_validation() {
    let r = Matrix::from_vec(3, 2, vec![0.01, 0.02, -0.01, 0.00, 0.03, 0.01]).unwrap();
    let s = sample_covariance(&r).unwrap();
    let e = ewma_covariance(&r, 0.999999).unwrap();
    assert!(max_abs_diff(&e, &s) < 1e-8);
    assert!(ewma_covariance(&r, 1.0).is_err());
    assert!(ewma_covariance(&r, 0.0).is_err());
}

#[test]
fn covariance_from_vols_known() {
    let vols = [0.01, 0.02];
    let corr = Matrix::from_vec(2, 2, vec![1.0, 0.5, 0.5, 1.0]).unwrap();
    let cov = covariance_from_vols(&vols, &corr).unwrap();
    assert_eq!(cov.get(0, 0), 1e-4);
    assert_eq!(cov.get(0, 1), 1e-4);
    assert_eq!(cov.get(1, 1), 4e-4);
}
